src/NuGet.Clients/PackageManagement.UI/Services/NuGetLockService.cs

```diff
@@ -18,8 +20,44 @@ public sealed class NuGetLockService : INuGetLockService, IDisposable
     {
         private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
 
+        private readonly JoinableTaskCollection _joinableCollection;
+        private readonly JoinableTaskFactory _joinableFactory;
+
         public bool IsLockHeld => _semaphore.CurrentCount == 0;
 
+        public NuGetLockService()
+        {
+            var joinableTaskContextNode = new JoinableTaskContextNode(ThreadHelper.JoinableTaskContext);
+            _joinableCollection = joinableTaskContextNode.CreateCollection();
+            _joinableFactory = joinableTaskContextNode.CreateFactory(_joinableCollection);
+        }
+
+        public JoinableTask<T> EnterNuGetOperation<T>(Func<System.Threading.Tasks.Task<T>> execute, CancellationToken token)
+        {
+            return _joinableFactory.RunAsync<T>(async delegate
+            {
+                using (_joinableCollection.Join())
+                {
+                    using (await AcquireLockAsync(token))
+                    {
+                        return await NuGetUIThreadHelper.JoinableTaskFactory.RunAsync<T>(async () =>
+                        {
+                            return await execute();
+                        });
+                    }
+                }
+            });
+        }
+
+        public JoinableTask EnterNuGetOperation(Func<System.Threading.Tasks.Task> execute, CancellationToken token)
+        {
+            return EnterNuGetOperation<bool>(async () =>
+            {
+                await execute();
+                return true;
+            }, token);
+        }
+
         public IDisposable AcquireLock()
         {
             _semaphore.Wait();
```

Comment:
This will almost certainly deadlock in many scenarios. I would recommend you delete this method entirely. But if you keep it, at least you should use JTF.Run and then call the async alternative method to await on the semaphore.